// ng-bootstrap-form-validation/bundles/ng-bootstrap-form-validation.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng-bootstrap-form-validation', ['exports', '@angular/core', '@angular/forms', '@angular/common'], factory) :
    (factory((global['ng-bootstrap-form-validation'] = {}),global.ng.core,global.ng.forms,global.ng.common));
}(this, (function (exports,i0,forms,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var FormValidationDirective = /** @class */ (function () {
        function FormValidationDirective() {
            this.validSubmit = new i0.EventEmitter();
        }
        /**
         * @return {?}
         */
        FormValidationDirective.prototype.onSubmit = /**
         * @return {?}
         */
            function () {
                this.markAsTouchedAndDirty(this.formGroup);
                if (this.formGroup.valid) {
                    this.validSubmit.emit(this.formGroup.value);
                }
            };
        /**
         * @param {?} control
         * @return {?}
         */
        FormValidationDirective.prototype.markAsTouchedAndDirty = /**
         * @param {?} control
         * @return {?}
         */
            function (control) {
                var _this = this;
                if (control instanceof forms.FormGroup) {
                    Object.keys(control.controls).forEach(function (key) {
                        return _this.markAsTouchedAndDirty(control.controls[key]);
                    });
                }
                else if (control instanceof forms.FormArray) {
                    control.controls.forEach(function (c) { return _this.markAsTouchedAndDirty(c); });
                }
                else if (control instanceof forms.FormControl && control.enabled) {
                    control.markAsDirty();
                    control.markAsTouched();
                    control.updateValueAndValidity();
                }
            };
        FormValidationDirective.decorators = [
            { type: i0.Directive, args: [{
                        // tslint:disable-next-line:directive-selector
                        selector: "[formGroup]"
                    },] },
        ];
        FormValidationDirective.propDecorators = {
            formGroup: [{ type: i0.Input }],
            validSubmit: [{ type: i0.Output }],
            onSubmit: [{ type: i0.HostListener, args: ["submit",] }]
        };
        return FormValidationDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /** @enum {number} */
    var BootstrapVersion = {
        Three: 0,
        Four: 1,
    };
    BootstrapVersion[BootstrapVersion.Three] = 'Three';
    BootstrapVersion[BootstrapVersion.Four] = 'Four';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /** @type {?} */
    var CUSTOM_ERROR_MESSAGES = new i0.InjectionToken("ng-bootstrap-form-validation custom error messages");
    /** @type {?} */
    var BOOTSTRAP_VERSION = new i0.InjectionToken("ng-bootstrap-form-validation module options");

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var MessagesComponent = /** @class */ (function () {
        function MessagesComponent(bootstrapVersion) {
            this.bootstrapVersion = bootstrapVersion;
            this.messages = function () { return []; };
        }
        Object.defineProperty(MessagesComponent.prototype, "className", {
            get: /**
             * @return {?}
             */ function () {
                switch (this.bootstrapVersion) {
                    case BootstrapVersion.Three:
                        return "help-block";
                    case BootstrapVersion.Four:
                        return "invalid-feedback";
                }
            },
            enumerable: true,
            configurable: true
        });
        MessagesComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: "bfv-messages",
                        template: "\n    <span [ngClass]=\"className\" *ngFor=\"let message of messages()\">{{message}}</span>\n  ",
                        styles: [
                            "\n      .invalid-feedback,\n      .valid-feedback {\n        display: block;\n      }\n    "
                        ]
                    },] },
        ];
        /** @nocollapse */
        MessagesComponent.ctorParameters = function () {
            return [
                { type: BootstrapVersion, decorators: [{ type: i0.Inject, args: [BOOTSTRAP_VERSION,] }] }
            ];
        };
        MessagesComponent.propDecorators = {
            messages: [{ type: i0.Input }]
        };
        return MessagesComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /** @type {?} */
    var DEFAULT_ERRORS = [
        {
            error: "required",
            format: function (label) { return label + " is required"; }
        },
        {
            error: "pattern",
            format: function (label) { return label + " is invalid"; }
        },
        {
            error: "minlength",
            format: function (label, error) {
                return label + " must be at least " + error.requiredLength + " characters";
            }
        },
        {
            error: "maxlength",
            format: function (label, error) {
                return label + " must be no longer than " + error.requiredLength + " characters";
            }
        },
        {
            error: "requiredTrue",
            format: function (label, error) { return label + " is required"; }
        },
        {
            error: "email",
            format: function (label, error) { return "Invalid email address"; }
        },
        {
            error: "max",
            format: function (label, error) { return label + " must be no greater than " + error.max; }
        },
        {
            error: "min",
            format: function (label, error) { return label + " must be no less than " + error.min; }
        }
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var ErrorMessageService = /** @class */ (function () {
        function ErrorMessageService(customErrorMessages) {
            this.customErrorMessages = customErrorMessages;
            this.defaultErrors = DEFAULT_ERRORS;
            this.errorMessages = customErrorMessages.reduce(function (acc, cur) { return acc.concat(cur); }, this.defaultErrors);
        }
        ErrorMessageService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: "root"
                    },] },
        ];
        /** @nocollapse */
        ErrorMessageService.ctorParameters = function () {
            return [
                { type: Array, decorators: [{ type: i0.Inject, args: [CUSTOM_ERROR_MESSAGES,] }] }
            ];
        };
        /** @nocollapse */ ErrorMessageService.ngInjectableDef = i0.defineInjectable({ factory: function ErrorMessageService_Factory() { return new ErrorMessageService(i0.inject(CUSTOM_ERROR_MESSAGES)); }, token: ErrorMessageService, providedIn: "root" });
        return ErrorMessageService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var FormGroupComponent = /** @class */ (function () {
        function FormGroupComponent(elRef, errorMessageService) {
            var _this = this;
            this.elRef = elRef;
            this.errorMessageService = errorMessageService;
            this.customErrorMessages = [];
            this.validationDisabled = false;
            this.messages = function () { return _this.getMessages(); };
        }
        Object.defineProperty(FormGroupComponent.prototype, "hasErrors", {
            get: /**
             * @return {?}
             */ function () {
                return (this.FormControlNames.some(function (c) { return !c.valid && c.dirty && c.touched; }) &&
                    !this.validationDisabled);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FormGroupComponent.prototype, "hasSuccess", {
            get: /**
             * @return {?}
             */ function () {
                return (!this.FormControlNames.some(function (c) { return !c.valid; }) &&
                    this.FormControlNames.some(function (c) { return c.dirty && c.touched; }) &&
                    !this.validationDisabled);
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        FormGroupComponent.prototype.ngAfterContentInit = /**
         * @return {?}
         */
            function () {
                if (this.messagesBlock) {
                    this.messagesBlock.messages = this.messages;
                }
            };
        /**
         * @return {?}
         */
        FormGroupComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.errorMessages = this.errorMessageService.errorMessages
                    .concat(this.customErrorMessages)
                    .reverse();
            };
        Object.defineProperty(FormGroupComponent.prototype, "label", {
            get: /**
             * @return {?}
             */ function () {
                /** @type {?} */
                var label = this.elRef.nativeElement.querySelector("label");
                return label && label.textContent ? label.textContent.trim() : "This field";
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FormGroupComponent.prototype, "isDirtyAndTouched", {
            get: /**
             * @return {?}
             */ function () {
                return this.FormControlNames.some(function (c) { return c.dirty && c.touched; });
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        FormGroupComponent.prototype.getMessages = /**
         * @return {?}
         */
            function () {
                var _this = this;
                /** @type {?} */
                var messages = [];
                if (!this.isDirtyAndTouched || this.validationDisabled) {
                    return messages;
                }
                /** @type {?} */
                var names = this.FormControlNames.map(function (f) { return f.name; });
                this.FormControlNames.filter(function (c, i) {
                    return !c.valid &&
                        !!c.errors &&
                        // filter out FormControlNames that share the same name - usually for radio buttons
                        names.indexOf(c.name) === i;
                }).forEach(function (control) {
                    Object.keys(control.errors).forEach(function (key) {
                        /** @type {?} */
                        var error = _this.errorMessages.find(function (err) { return err.error === key; });
                        if (!error) {
                            return;
                        }
                        messages.push(error.format(_this.label, control.errors[key]));
                    });
                });
                return messages;
            };
        FormGroupComponent.decorators = [
            { type: i0.Component, args: [{
                        // tslint:disable:component-selector
                        selector: ".form-group",
                        template: "\n    <ng-content></ng-content>\n    <bfv-messages *ngIf=\"!messagesBlock\" [messages]=\"messages\"></bfv-messages>\n  "
                    },] },
        ];
        /** @nocollapse */
        FormGroupComponent.ctorParameters = function () {
            return [
                { type: i0.ElementRef },
                { type: ErrorMessageService }
            ];
        };
        FormGroupComponent.propDecorators = {
            FormControlNames: [{ type: i0.ContentChildren, args: [forms.FormControlName,] }],
            customErrorMessages: [{ type: i0.Input }],
            validationDisabled: [{ type: i0.Input }],
            hasErrors: [{ type: i0.HostBinding, args: ["class.has-error",] }],
            hasSuccess: [{ type: i0.HostBinding, args: ["class.has-success",] }],
            messagesBlock: [{ type: i0.ContentChild, args: [MessagesComponent,] }]
        };
        return FormGroupComponent;
    }());

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /**
     * @param {?} name
     * @param {?} parent
     * @return {?}
     */
    function controlPath(name, parent) {
        // tslint:disable-next-line:no-non-null-assertion
        return __spread(/** @type {?} */ ((parent.path)), [name]);
    }
    var FormControlDirective = /** @class */ (function () {
        function FormControlDirective(
        // this value might be null, but we union type it as such until
        // this issue is resolved: https://github.com/angular/angular/issues/25544
        parent, bootstrapVersion) {
            this.parent = parent;
            this.bootstrapVersion = bootstrapVersion;
        }
        Object.defineProperty(FormControlDirective.prototype, "validClass", {
            get: /**
             * @return {?}
             */ function () {
                if (!this.control) {
                    return false;
                }
                return (this.bootstrapFour &&
                    this.control.valid &&
                    (this.control.touched || this.control.dirty));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FormControlDirective.prototype, "invalidClass", {
            get: /**
             * @return {?}
             */ function () {
                if (!this.control) {
                    return false;
                }
                return (this.bootstrapFour &&
                    this.control.invalid &&
                    this.control.touched &&
                    this.control.dirty);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FormControlDirective.prototype, "path", {
            get: /**
             * @return {?}
             */ function () {
                return controlPath(this.formControlName, this.parent);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FormControlDirective.prototype, "control", {
            get: /**
             * @return {?}
             */ function () {
                return this.formDirective && this.formDirective.getControl(this);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FormControlDirective.prototype, "formDirective", {
            get: /**
             * @return {?}
             */ function () {
                return this.parent ? this.parent.formDirective : null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FormControlDirective.prototype, "bootstrapFour", {
            get: /**
             * @return {?}
             */ function () {
                return this.bootstrapVersion === BootstrapVersion.Four;
            },
            enumerable: true,
            configurable: true
        });
        FormControlDirective.decorators = [
            { type: i0.Directive, args: [{
                        // tslint:disable-next-line:directive-selector
                        selector: ".form-control,.form-check-input,.custom-control-input"
                    },] },
        ];
        /** @nocollapse */
        FormControlDirective.ctorParameters = function () {
            return [
                { type: forms.ControlContainer, decorators: [{ type: i0.Optional }, { type: i0.Host }, { type: i0.SkipSelf }] },
                { type: BootstrapVersion, decorators: [{ type: i0.Inject, args: [BOOTSTRAP_VERSION,] }] }
            ];
        };
        FormControlDirective.propDecorators = {
            formControlName: [{ type: i0.Input }],
            formControl: [{ type: i0.Input }],
            validClass: [{ type: i0.HostBinding, args: ["class.is-valid",] }],
            invalidClass: [{ type: i0.HostBinding, args: ["class.is-invalid",] }]
        };
        return FormControlDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgBootstrapFormValidationModule = /** @class */ (function () {
        function NgBootstrapFormValidationModule() {
        }
        /**
         * @param {?=} userOptions
         * @return {?}
         */
        NgBootstrapFormValidationModule.forRoot = /**
         * @param {?=} userOptions
         * @return {?}
         */
            function (userOptions) {
                if (userOptions === void 0) {
                    userOptions = {
                        bootstrapVersion: BootstrapVersion.Four
                    };
                }
                return {
                    ngModule: NgBootstrapFormValidationModule,
                    providers: [
                        {
                            provide: CUSTOM_ERROR_MESSAGES,
                            useValue: userOptions.customErrorMessages || [],
                            multi: true
                        },
                        {
                            provide: BOOTSTRAP_VERSION,
                            useValue: userOptions.bootstrapVersion
                        },
                        ErrorMessageService
                    ]
                };
            };
        NgBootstrapFormValidationModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [common.CommonModule],
                        declarations: [
                            FormValidationDirective,
                            FormGroupComponent,
                            MessagesComponent,
                            FormControlDirective
                        ],
                        exports: [
                            FormValidationDirective,
                            FormGroupComponent,
                            MessagesComponent,
                            FormControlDirective
                        ]
                    },] },
        ];
        return NgBootstrapFormValidationModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.FormValidationDirective = FormValidationDirective;
    exports.MessagesComponent = MessagesComponent;
    exports.FormGroupComponent = FormGroupComponent;
    exports.BootstrapVersion = BootstrapVersion;
    exports.NgBootstrapFormValidationModule = NgBootstrapFormValidationModule;
    exports.CUSTOM_ERROR_MESSAGES = CUSTOM_ERROR_MESSAGES;
    exports.BOOTSTRAP_VERSION = BOOTSTRAP_VERSION;
    exports.ɵb = FormControlDirective;
    exports.ɵa = ErrorMessageService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
